### PR TITLE
[MLIR] Fix crash in AffineMap::replace for zero result maps

### PR DESCRIPTION
**Pull request by bondhugula**
Fix obvious bug in AffineMap::replace for the case of zero result maps.
Extend/complete inferExprsFromList to work with empty expression lists.

### CODE REVIEW

**Comment by bondhugula** [comment]
The other option here is to leave the current behavior of asserting in `inferFromExprsList` on empty lists (which isn't btw specified/documented on the API doc comment) and instead update AffineMap::replace to handle the zero result case with an early return, and document the non-empty expectation on `inferFromExprsList`. But I felt removing those asserts was more systematic and complete.

File: mlir/include/mlir/IR/AffineMap.h [hunk 1]
```diff
@@ -122,9 +122,11 @@ class AffineMap {
   /// `exprs.size()`, as many dims as the largest dim in `exprs` and as many
   /// symbols as the largest symbol in `exprs`.
   static SmallVector<AffineMap, 4>
-  inferFromExprList(ArrayRef<ArrayRef<AffineExpr>> exprsList);
+  inferFromExprList(ArrayRef<ArrayRef<AffineExpr>> exprsList,
+                    MLIRContext *context);
   static SmallVector<AffineMap, 4>
-  inferFromExprList(ArrayRef<SmallVector<AffineExpr, 4>> exprsList);
+  inferFromExprList(ArrayRef<SmallVector<AffineExpr, 4>> exprsList,
+                    MLIRContext *context);
```
Comment (joker-eph) [inline mlir/include/mlir/IR/AffineMap.h:129]: Nit: I think most APIs take the context as first argument? (thinking of all the attribute / types creation for example)

**Review by joker-eph — APPROVED** [review]

File: mlir/include/mlir/IR/AffineMap.h (hunk 1 above)
Comment (bondhugula) [reply mlir/include/mlir/IR/AffineMap.h:129]: Unfortunately, the codebase isn't following a convention here, and things are being done both ways. Here's a crude search result from `llvm/mlir/`:
```
$ grep  ", context" lib/ -r | wc -l
216
grep  "(context, " lib/ -r | wc -l
247
$ grep  ", context)" lib/IR -r | wc -l
60
$ grep  "(context, " lib/IR -r | wc -l
34
```
From a usage standpoint, I think having it as a trailing argument fits my mental model better, going by "important things come first." So, thinking of the key inputs when opening the parenthesis is natural, followed by the other baggage.

**Review by matthias-springer — APPROVED** [review]

**Review by nicolasvasilache — DISMISSED** [review]
I'd supersede the "important arguments" point by the "default arguments".
Do we expect this to support default arguments or not?
If not this is fine, otherwise we need to put the mandatory args first.

Putting a block to get an answer to this (didn't look at the code here).

**Comment by bondhugula** [comment]
> I'd supersede the "important arguments" point by the "default arguments". Do we expect this to support default arguments or not? If not this is fine, otherwise we need to put the mandatory args first.
> 
> Putting a block to get an answer to this (didn't look at the code here).

There are no default arguments here now or later that I can think of.

**Pull request merged**